### DISCUSSION TITLE
Stacking Dynamic outputs

### DISCUSSION THREAD

**Discussion by dmsfabiano** [dagster-io/dagster#15630] [Q&A] [answered]
Can you chain Dynamic Outputs in Dagster? say I have an op that fetches the batch size for my ML runs, it returns 5 objects, and for each of those objects I have to run N tasks,

I created an op that yields each object individually as an DynamicOutput but I cannot find any examples online on how to pass those as inputs to another dynamic output generator. Any pointers to documentation?

Here is what I am trying to do, but it is giving me an invalid definition error

```python
@op(out=DynamicOut())
def get_ml_job_batch() -> List[MLJob]:  # noqa: D103
    batch_size = 5
    session = Session(engine)
    ml_jobs = (
        session.query(MLJob)
        .filte(MLJob.status == Status.STAGED)
        .order_by(MLJob.id.asc())
        .limit(batch_size)
    )
    session.close()
    for idx, ml_job in enumerate(ml_jobs):
        yield DynamicOutput(ml_job, mapping_key=idx)

@op(out=DynamicOut())
def get_ml_job_tasks(ml_job) -> List[MLTask]:  # noqa: D103
    session = Session(engine)
    tasks = (
        session.query(MLTask).filter(MLTask.parent_id == ml_job.id).order_by(MLTask.id.asc())
    )
    session.close()
    for idx, task in enumerate(tasks):
        yield DynamicOutput(task, mapping_key=idx)

@op
def compute_extraction(task):
    collection_json = json.loads(Path(task.path).read_text())
    api_path = http:///models/brazil_sanctions/predict  # noqa: E501
    session = Session(engine)
    extraction = requests.get(api_path, json=collection_json)

    task.status = Status.DONE
    session.commit()
    session.close()

@job
def execute_ml_jobs():
    session = Session(engine)
    jobs = get_ml_job_batch().collect()
    for _job in jobs:
        get_ml_job_tasks(_job).map(compute_extraction)
        _job.status = Status.DONE
    session.commit()
```

Stack trace:
```
Error loading repository location citiflow.pocs.orchestrator.dagster:dagster._core.errors.DagsterInvalidDefinitionError: In @job execute_ml_jobs, received invalid type <class 'str'> for input "ml_job" (at position 0) in op invocation "get_ml_job_tasks". Must pass the output from previous node invocations or inputs to the composition function as inputs when invoking nodes during composition.
```

**Comment by dpeng817** [comment, marked as answer]
Hey there! 
I think the main issue here is that you are decomposing the `jobs` variable retrieved from the call to `get_ml_job_batch().collect()`. The `@job` dsl just encodes dependency information, and expects that you pass the collected result directly to the next op. So instead, you could do:
```python
@job
def execute_ml_jobs():
    ...
    jobs = get_ml_job_batch().collect()
    get_ml_job_tasks(jobs).map(compute_extraction)
    ...
```

And `get_ml_job_tasks` would take in a list of results, as described here:
```python
@op(out=DynamicOut())
def get_ml_job_tasks(ml_jobs) -> List[MLTask]:  # noqa: D103
    session = Session(engine)
    for ml_job in ml_jobs:
        tasks = (
            session.query(MLTask).filter(MLTask.parent_id == ml_job.id).order_by(MLTask.id.asc())
        )
        session.close()
        for idx, task in enumerate(tasks):
            yield DynamicOutput(task, mapping_key=(ml_job.id, idx))
```
Not sure what the best mapping key situation would be here, what I provided is just an example and depends on your actual use case.

A few additional thoughts:
- The `@job` decorator just builds out dependency information, and runs immediately. I don't know the context of the call to `Session`, but it won't be declaring a global session for the duration of the run if that's the intention. Recommend using a resource for that
- Dagster uses the type hints you provide for runtime type checking - I don't think it's causing your current problem, but will likely cause problems down the road, since those annotations don't match the function body, from what I can see.
Hope that helps clear things up.
